flake.nix

```diff
@@ -6,46 +6,91 @@
+        cargoPackageVersion = cargoToml.package.version;
 
-        buildXwaylandSatellite = { withSystemd ? false }: naersk'.buildPackage {
-          src = ./.;
+        commitHash = self.shortRev or self.dirtyShortRev or "unknown";
 
-          nativeBuildInputs = with pkgs; [
-            rustPlatform.bindgenHook
-            rust-bin.stable.latest.default
-            pkg-config
+        version = "${cargoPackageVersion}-${commitHash}";
 
-            xcb-util-cursor
-            xorg.libxcb
+        buildXwaylandSatellite =
+          { lib
+          , rustPlatform
+          , pkg-config
+          , makeBinaryWrapper
+          , libxcb
+          , xcb-util-cursor
+          , xwayland
+          , withSystemd ? true
+          }:
 
-            makeWrapper
-          ] ++ lib.optional withSystemd pkgs.systemd;
+          rustPlatform.buildRustPackage rec {
+            pname = "xwayland-satellite";
+            inherit version;
 
-          buildInputs = [ pkgs.xwayland ];
+            src = self;
 
-          cargoBuildOptions = opts: opts ++ lib.optional withSystemd "--features systemd";
+            cargoLock = {
+              lockFile = "${src}/Cargo.lock";
+              allowBuiltinFetchGit = true;
+            };
 
-          postInstall = ''
-            wrapProgram $out/bin/xwayland-satellite \
-              --prefix PATH : ${pkgs.xwayland}/bin
-          '';
-        };
+            nativeBuildInputs = [
+              rustPlatform.bindgenHook
+              pkg-config
+              makeBinaryWrapper
+            ];
+
+            buildInputs = [
+              libxcb
+              xcb-util-cursor
+            ];
+
+            buildNoDefaultFeatures = true;
+            buildFeatures = lib.optionals withSystemd [ "systemd" ];
 
+            postPatch = ''
+              substituteInPlace resources/xwayland-satellite.service \
+                --replace-fail '/usr/local/bin' "$out/bin"
+            '';
+
+            postInstall = lib.optionalString withSystemd ''
+              install -Dm0644 resources/xwayland-satellite.service -t $out/lib/systemd/user
+            '';
+
+            postFixup = ''
+              wrapProgram $out/bin/xwayland-satellite \
+                --prefix PATH : "${lib.makeBinPath [ xwayland ]}"
+            '';
+
+            doCheck = false;
+
+            meta = with lib; {
+              description = "Xwayland outside your Wayland";
+              homepage = "https://github.com/Supreeeme/xwayland-satellite";
+              license = licenses.mpl20;
+              platforms = platforms.linux;
+            };
+          };
+
+        xwayland-satellite = pkgs.callPackage buildXwaylandSatellite {
+          lib = pkgs.lib;
+          rustPlatform = pkgs.rustPlatform;
+          pkg-config = pkgs.pkg-config;
+          makeBinaryWrapper = pkgs.makeBinaryWrapper;
+          libxcb = pkgs.xorg.libxcb;
+          xcb-util-cursor = pkgs.xcb-util-cursor;
+          xwayland = pkgs.xwayland;
```

Comment:
`pkgs.callPackage` prepopulates this. that's the whole point of `pkgs.callPackage`; it's `makeOverridable` but with any attribute from nixpkgs already filled in.

```suggestion
```

`libxcb` won't be filled in automatically though, i think. in that case, the closure should take `xorg` instead, and internally use `xorg.libxcb`.

also, is `libxcb` necessary? other packages (niri-flake, nixpkgs) don't mention it. is some functionality broken there as a result? cc @Supreeeme

Reply:
other than being more concise, there is a subtle difference with `callPackage`: it will pass the correct versions of dependencies for *cross-compilation*. i guess we don't expose cross-comp here? to make use of this, the `pkgs` in `pkgs.callPackage` must be made for cross-compilation. and here, the `pkgs` definitely isn't cross-compilable. but in general, when the buiild and target platform are not identical, then manually specifying the dependencies is actually going to cause problems because `pkgs.*` is for the target platform, but e.g. `rustPlatform` should come from the build platform.

Reply:
> `pkgs.callPackage` prepopulates this.

Addressed in 8c56440.

> other than being more concise, there is a subtle difference with `callPackage`: it will pass the correct versions of dependencies for _cross-compilation_.

Got it.

> also, is `libxcb` necessary? other packages (niri-flake, nixpkgs) don't mention it.

`libxcb` is indeed included as a dependency in `xwayland-satellite`'s nixpkgs [expression](https://github.com/NixOS/nixpkgs/blob/7eee17a8a5868ecf596bbb8c8beb527253ea8f4d/pkgs/by-name/xw/xwayland-satellite/package.nix#L32).

Reply:
re: libxcb, the library used for the x11 protocol is in fact a rust wrapper around libxcb 🙂

Reply:
> `libxcb` is indeed included as a dependency in `xwayland-satellite`'s nixpkgs [expression](https://github.com/NixOS/nixpkgs/blob/7eee17a8a5868ecf596bbb8c8beb527253ea8f4d/pkgs/by-name/xw/xwayland-satellite/package.nix#L32).

oh. i missed that. [i don't have it in `niri-flake`](https://github.com/sodiboo/niri-flake/blob/d7c058c0c0b8cd5b799a3a3ac4f8a1090454155e/flake.nix#L228-L229), is that package broken as a result then?

but yeah, seems i was wrong. this is probably necessary.